### DISCUSSION TITLE
ConnectionOpen -> ConnectionStart - question about ports in use

### DISCUSSION THREAD

**Discussion by timprepscius** [microsoft/msquic#1701] [General]
Am I allowed to:
(pseudo)

var connection = connectionOpen();
connectionStart(connection, ..., callbackForServerA, serverA_IP, serverAPort);
connectionStart(connection, ..., callbackForServerB, serverB_IP, serverBPort);

Is this allowed?  If I want to make connections to let's say, 10000 computers,
do I need to use more than one port ?

When is the port actually created (I assumed in connectionOpen)?

---

In a nutshell, I am looking for a solution where a client uses only one UDP port to connect to multiple servers.
Just as the server would use one port to listen for connections. 

Is this possible with QUIC?

Thanks,

-tim

**Comment by nibanks** [comment]
You can use a single port to practically start unlimited client connections. You do have to open separate connection objects/handles for each. The one extra step you need to set `QUIC_PARAM_CONN_SHARE_UDP_BINDING` to `TRUE` before calling start.

**Reply by timprepscius** [reply]
Thanks for this.

**Reply by timprepscius** [reply]
I have probably a super obvious question:

So, I'm reading through RpsClient.cpp

So let's say I:
1. connectionOpen
with my connectionHandler
2. and then set the flag of the QUIC_PARAM_CONN_SHARE_UDP_BINDING

and then I do a 
3. connectionStart to serverA
4. and connectionStart to serverB


Let's say serverB initiates a shutdown
SHUTDOWN_INITIATED_BY_PEER

I assume my connectionHandler will receive that event.
But how do I know whether it pertains to serverA or serverB?
I'm looking at msquic.h: QUIC_CONNECTION_EVENT
and I'm not seeing how to differentiate events for different connectionStarts


---
Or am I getting this completely wrong?

Thanks,

-tim

**Reply by ThadHouse** [reply]
The HQUIC will be different, and each connection can also be passed its own context pointer that you can use to differentiate between the connections.

**Reply by timprepscius** [reply]
Oh, I reread nibanks comment, and it says, "You do have to open separate connection objects/handles for each."

Ah, so I am doing another open before each connectionStart.  Sorry, brain skipped a few words.
